torch/_subclasses/fake_tensor.py

```diff
@@ -292,6 +222,99 @@ def __init__(self):
         self.fake_tensor_converter = FakeTensorConverter()
 
     def __torch_dispatch__(self, func, types, args=(), kwargs=None):
-        def run_fn(func, types, args, kwargs):
-            return func(*args, **kwargs)
-        return torch_dispatch_impl(self, func, types, args, kwargs, run_fn)
+        kwargs = kwargs if kwargs else {}
+
+        # TODO: apply as no_dispatch decorator
+        with no_dispatch():
```

Comment:
Would be great to have a test suite/ demo / something to test if your new tensor subclass is compositional because it's a bit hard to reason about without actually going through tests/debugging/seeing how things work together. In subclass zoo for example

Reply:
Without the no-dispatch I get infinite recursion. the `r = func(*args, **kwargs)` [here](https://github.com/pytorch/pytorch/pull/78677/files#diff-39e82af71afdadbc56a4ecf552ec668ddcda794f8ea3ec41edaef23456fd56e9R295) just calls into `__torch_dispatch__` again

Reply:
If we want this to work and it to be compositional with other modes, we'll need to unwrap args and kwargs. Usually this would just be `arg.elem if isinstance(arg, FakeTensor) else arg` but it seems like this will probably be more complicated before it can be a wrapper tensor? Not sure if @ezyang has any thoughts on how to do this

Reply:
Ah looks like this should work--so basically we should be saving elem and then unwrapping to that elem, like Ed said, if the mode matches your mode (currently this should always happen since we error if there's FakeTensors from more than one mode but we can be extra careful to set us up for nested FakeTensors if we want)

I got confused by [this comment](https://github.com/pytorch/pytorch/pull/78677/files#diff-39e82af71afdadbc56a4ecf552ec668ddcda794f8ea3ec41edaef23456fd56e9R125) though: `elem does not need to be recorded, because FakeTensor *is a* elem`. FakeTensor should only be taking a vanilla tensor as elem?

Reply:
Ah I see--that is Tracer/ProxyTensor specific. And kinda quirky 

Yeah so the idea is that this is basically `FakeTensor(torch.Tensor(...))` and we want to redispatch on `torch.Tensor` instead of on FakeTensor so that we don't get the infinite recursion you're seeing

So if we save `elem`, we should be able to unwrap by basically doing `arg.elem if isinstance(arg, FakeTensor) and self == arg.fake_mode else arg` for every arg and kwarg (this adds in Ed's check that we only unwrap tensors in the same mode) and then run `func(*new_args, **new_kwargs)`

Reply:
There is no `elem`, since FakeTensor inherits it's not compositional.. im going to defer this working for another pr